chain/consensus/mir/crypto.go

```diff
@@ -0,0 +1,124 @@
+package mir
+
+import (
+	"context"
+	"crypto/sha256"
+	"strings"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	filcrypto "github.com/filecoin-project/go-state-types/crypto"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/lib/sigs"
+	mir "github.com/filecoin-project/mir/pkg/modules"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+var MsgMeta = api.MsgMeta{Type: "mir-request"}
+
+var _ mir.Crypto = &CryptoManager{}
+
+type CryptoManager struct {
+	addr   address.Address
+	wallet api.Wallet
+}
+
+func NewCryptoManager(addr address.Address, wallet api.Wallet) (*CryptoManager, error) {
+	if addr.Protocol() != address.SECP256K1 {
+		return nil, xerrors.New("must be SECP address")
+	}
+	return &CryptoManager{addr, wallet}, nil
+}
+
+// Sign signs the provided data and returns the resulting signature.
+// The data to be signed is the concatenation of all the passed byte slices.
+// A signature produced by Sign is verifiable using VerifyNodeSig or VerifyClientSig,
+// if, respectively, RegisterNodeKey or RegisterClientKey has been invoked with the corresponding public key.
+// Note that the private key used to produce the signature cannot be set ("registered") through this interface.
+// Storing and using the private key is completely implementation-dependent.
+func (c *CryptoManager) Sign(data [][]byte) (bytes []byte, err error) {
+	signature, err := c.wallet.WalletSign(context.Background(), c.addr, hash(data), MsgMeta)
+	return signature.MarshalBinary()
+}
+
+// VerifyNodeSig verifies a signature produced by the node with numeric ID nodeID over data.
+// Returns nil on success (i.e., if the given signature is valid) and a non-nil error otherwise.
+// Note that RegisterNodeKey must be used to register the node's public key before calling VerifyNodeSig,
+// otherwise VerifyNodeSig will fail.
+func (c *CryptoManager) VerifyNodeSig(data [][]byte, signature []byte, nodeID t.NodeID) error {
+	nodeAddr, err := getAddr(nodeID.Pb())
+	if err != nil {
+		return err
+	}
+	return c.verifySig(data, signature, nodeAddr)
+}
+
+// RegisterNodeKey associates a public key with a numeric node ID.
+// The representation of the key is implementation-dependent.
+// Calls to VerifyNodeSig will fail until RegisterNodeKey is successfully called with the corresponding node ID.
+// Returns nil on success, a non-nil error on failure.
+func (c *CryptoManager) RegisterNodeKey(pubKey []byte, nodeID t.NodeID) error {
+	panic("not implemented")
```

Comment:
Maybe I'm missing something but, why do we need to sign with the host identity and with a `walletID`? We should have look at https://github.com/libp2p/go-libp2p-core/tree/master/crypto, libp2p hosts already hold an identity and can sign with their libp2p identity.

Reply:
In fact, the `RegisterNodeKey` (and other functions importing and deleting keys) are gone anyway in the newest restructured Mir implementation. It's really just about signing and verifying now.